File: app/src/main/java/com/teamunemployment/lolanalytics/front_page/Tabs/MatchHistoryTab/MatchHistoryInteractor.java
package com.teamunemployment.lolanalytics.front_page.Tabs.MatchHistoryTab;

import android.content.Context;

import com.teamunemployment.lolanalytics.io.RESTApiExecutor;
import com.teamunemployment.lolanalytics.io.RealmExecutor;
import com.teamunemployment.lolanalytics.data.model.MatchHistoryData;
import com.teamunemployment.lolanalytics.data.model.MatchIdWrapper;
import com.teamunemployment.lolanalytics.data.model.MatchSummary;
import com.teamunemployment.lolanalytics.front_page.Tabs.MatchHistoryTab.Cards.MatchHistoryCardViewContract;
import com.teamunemployment.lolanalytics.front_page.Tabs.MatchHistoryTab.Model.MatchHistoryCardData;
import com.teamunemployment.lolanalytics.Utils.Network;

import java.util.ArrayList;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import io.realm.Realm;

/**
 * TODO refactor this class to not have hardcoded observers.
 * @author Josiah Kendall
 */
public class MatchHistoryInteractor {
//    Network network = new Network();

    private RESTApiExecutor restApiExecutor;
    private RealmExecutor realmExecutor;
    private Context context;

    public MatchHistoryInteractor(RESTApiExecutor restApiExecutor, RealmExecutor realmExecutor,
                                  Context context) {
        this.restApiExecutor = restApiExecutor;
        this.realmExecutor = realmExecutor;
        this.context = context;
    }

    public void LoadCachedMatchHistoryData(final int role, final long summonerId, final MatchHistoryTabContract.Presenter presenter) {
        // Create an observable for fetching our cached value, as we want to get this off of the base thread.
        Observable<ArrayList<MatchIdWrapper>> observable = Observable.create((ObservableEmitter<ArrayList<MatchIdWrapper>> emitter) -> {
            Realm.init(context);
            Realm realm = Realm.getDefaultInstance();
            ArrayList<MatchIdWrapper> data = realmExecutor.LoadMatchList(realm, summonerId);
            emitter.onNext(data);
            emitter.onComplete();
        });

        observable.subscribeOn(AndroidSchedulers.mainThread()) // Gotta be on the main thread for realm. If performance is an issue, use fetchAsync()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<ArrayList<MatchIdWrapper>>() {
                    @Override
                    public void onError(Throwable e) {
                        presenter.onError(e);
                    }

                    @Override
                    public void onComplete() {

                    }

                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(ArrayList<MatchIdWrapper> matchHistoryData) {
                        presenter.onMatchListLoadedSuccessfully(matchHistoryData);
                    }
                });
    }

    /**
     * Fetch our match history value from the server.
     * @param role The role we want the value for.
     * @param summonerId .
     * @param presenter The presenter to return the value to.
     */
    public void LoadFreshMatchHistoryData(int role, long summonerId, final MatchHistoryPresenter presenter) {
//        if (!network.isConnectingToInternet(context)) {
//            return;
//        }
        Observable<MatchHistoryData> matchHistoryDataObservable = restApiExecutor.GetMatchListForSummonerInSpecificRole(summonerId, role);
        matchHistoryDataObservable
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.computation())
                .map(matchHistoryData -> {
                    Realm.init(context);
                    Realm realm = Realm.getDefaultInstance();
                    realmExecutor.SaveMatchList(realm, matchHistoryData.getItems());
                    return matchHistoryData.getItems();
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<ArrayList<MatchIdWrapper>>() {

                    @Override
                    public void onError(Throwable e) {
                        presenter.onError(e);
                    }

                    @Override
                    public void onComplete() {

                    }

                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(ArrayList<MatchIdWrapper> matchHistoryData) {
                        presenter.onMatchListLoadedSuccessfully(matchHistoryData);
                    }
                });
    }

    // Think these two methods might be useless.
    public Observable<MatchSummary> LoadFreshMatchSummary(long matchId, MatchHistoryPresenter matchHistoryPresenter) {
        return null;
    }

    public Observable<MatchSummary> LoadCachedMatchSummary(long matchId, MatchHistoryPresenter matchHistoryPresenter) {
        return null;
    }

    /**
     * Load the match details to show on a card.
     * @param id The id of the match as per saved in our database. may or may not be the same as the Riot match id.
     * @param presenter The presenter to return the value to when we are done loading.
     */
    public void LoadMatchDetails(long id,
                                 MatchHistoryTabContract.Presenter presenter,
                                 MatchHistoryCardViewContract cardViewContract) {
        // Fetch from the cache.
        Realm.init(context);
        Realm realm = Realm.getDefaultInstance();
        MatchHistoryCardData matchHistoryCardData = realmExecutor.GetSingleMatchHistoryCard(realm, id);
        if (matchHistoryCardData != null) {
            cardViewContract.setChampName(matchHistoryCardData.getChampName());
            cardViewContract.setGraph1(matchHistoryCardData.getKills());
            cardViewContract.setGraph2(matchHistoryCardData.getDeaths());
            cardViewContract.setGraph3(matchHistoryCardData.getCsTotal());
            cardViewContract.setKDA("3/2/5"); // TODO
        } else {
            // If there is nothing there, fetch from the server.
//            if (network.isConnectingToInternet(context)) {
                fetchFreshMatchDetails(id, presenter, cardViewContract);
//            }
        }
    }

    /**
     * THis should only ever be fetched once, as match details should never change.
     * @param id
     * @param presenter
     * @param cardViewContract
     */
    private void fetchFreshMatchDetails(final long id, MatchHistoryTabContract.Presenter presenter, MatchHistoryCardViewContract cardViewContract) {
        Observable<MatchHistoryCardData> matchHistoryCardDataObservable = restApiExecutor.GetMatchHistoryCardData(id, -1); // TODO
        matchHistoryCardDataObservable
                .subscribeOn(Schedulers.newThread())
                .observeOn(Schedulers.computation())
                .map(matchHistoryCardData -> {
                    Realm.init(context);
                    Realm realm = Realm.getDefaultInstance();

                    matchHistoryCardData.setMatchId(id);
                    realmExecutor.SaveMatchHistoryCardData(realm, matchHistoryCardData);
                    return matchHistoryCardData;
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<MatchHistoryCardData>() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onNext(MatchHistoryCardData value) {
                presenter.SetLoadedCardData(value, cardViewContract);
            }

            @Override
            public void onError(Throwable e) {
                // TODO
                presenter.handleError(e);
            }

            @Override
            public void onComplete() {

            }
        });
    }
}

File: app/src/main/java/com/teamunemployment/lolanalytics/front_page/tabs/MatchHistoryTab/MatchHistoryInteractor.kt
package com.teamunemployment.lolanalytics.front_page.tabs.MatchHistoryTab

import android.content.Context
import android.util.Log
import co.metalab.asyncawait.async
import com.teamunemployment.lolanalytics.Utils.Network
import com.teamunemployment.lolanalytics.Utils.Role
import com.teamunemployment.lolanalytics.Utils.RoleUtils
import com.teamunemployment.lolanalytics.Utils.getMatchHistoryCardData
import com.teamunemployment.lolanalytics.data.model.Result
import com.teamunemployment.lolanalytics.data.room.MatchHistory.MatchHistoryDao
import com.teamunemployment.lolanalytics.data.room.champ.ChampDao

import com.teamunemployment.lolanalytics.front_page.tabs.MatchHistoryTab.Cards.MatchHistoryCardViewContract
import com.teamunemployment.lolanalytics.front_page.tabs.MatchHistoryTab.Model.MatchHistoryCardData
import com.teamunemployment.lolanalytics.front_page.tabs.MatchHistoryTab.Model.MatchIdentifier
import com.teamunemployment.lolanalytics.io.networking.RetrofitFactory
import retrofit2.Call
import retrofit2.Response
import ru.gildor.coroutines.retrofit.await
import kotlin.collections.ArrayList

/**
 * TODO refactor this class to not have hardcoded observers.
 * @author Josiah Kendall
 */
class MatchHistoryInteractor(private val context: Context,
                             private val retrofitFactory: RetrofitFactory,
                             private val network: Network,
                             private val matchHistoryDao: MatchHistoryDao,
                             private val champDao: ChampDao) {
    private val TAG = "MatchHistoryInteractor"
    private val roleUtils = Role()
    /**
     * Load the match details to show on a card.
     * @param id The id of the match as per saved in our database. may or may not be the same as the Riot match id.
     * @param presenter The presenter to return the value to when we are done loading.
     */
    fun loadMatchDetails(id: Long,
                         role : Int,
                         presenter : MatchHistoryTabContract.Presenter,
                         cardViewContract : MatchHistoryCardViewContract,
                         region: String,
                         summonerId : String,
                         timestamp : Long) {
        async {
            val url = network.getUrlForRegion(region)
            val matchHistoryService = retrofitFactory.produceRetrofitInterface(MatchHistoryService::class.java, url)
            val call: Call<Result<MatchHistoryCardData>> = matchHistoryService.fetchMatchSummary(role,id,summonerId)
            Log.d(TAG, "Fetching match card with role $role and id $id and summonerId $summonerId")

            val response: Response<Result<MatchHistoryCardData>> = await { call.execute() }
            Log.d(TAG, "Result was : ${response.code()}")
            Log.v(TAG, "Body was : ${response.body()}")

            val result: Result<MatchHistoryCardData> = response.getMatchHistoryCardData()
            val code = result.resultCode
            if (code ==200) {
                Log.d(TAG, "Found data : ${result.data}")
                await { matchHistoryDao.insert(result.data) }
                val champ = await { champDao.loadChamp(result.data.champId.toString(), summonerId) }
                if (champ != null) {
                    presenter.setLoadedCardData(result.data, cardViewContract, timestamp, "${produceResult(result.data.won)} ${champ.title}" )
                } else {
                    presenter.setLoadedCardData(result.data, cardViewContract, timestamp, "")
                }
            }
        }.onError {
            presenter.showMessage("Offline - check you internet connection")
        }
    }

    private fun produceResult(win : Boolean) : String {
        return if (win) {
            "Win as "
        } else "Loss as "
    }

    fun loadMatches(region: String,
                    summonerId : String,
                    role: Int,
                    startingPoint: Int,
                    presenter: MatchHistoryPresenter) {
        val laneString = roleUtils.produceCorrectLaneStringForRoleInt(role)
        val roleString = roleUtils.produceCorrectRoleStringForRoleInt(role)
        async {
            // try fetch from cache initially

            val matchIdentifierList = await{ matchHistoryDao.loadTwentyMatchIdenfifiers(summonerId, roleString, laneString, startingPoint)}
            if (matchIdentifierList.isEmpty()) {
                val url = network.getUrlForRegion(region)
                val matchHistoryService = retrofitFactory.produceRetrofitInterface(MatchHistoryService::class.java, url)
                val result: Result<ArrayList<MatchIdentifier>> = await{matchHistoryService.fetchMatches(role, startingPoint,summonerId).execute().body()!!}
                await{cacheMatchList(result.data)}
                presenter.onMatchListLoadedSuccessfully(result.data, result.resultCode)
            } else {
                presenter.onMatchListLoadedSuccessfully(matchIdentifierList as java.util.ArrayList<MatchIdentifier>,200)
            }
        }.onError {
            presenter.showMessage("Offline - check you internet connection")
        }
    }

    fun loadMatches(region: String,
                    summonerId : String,
                    role: Int,
                    champId : Int,
                    startingPoint: Int,
                    presenter: MatchHistoryPresenter) {
        val laneString = roleUtils.produceCorrectLaneStringForRoleInt(role)
        val roleString = roleUtils.produceCorrectRoleStringForRoleInt(role)
        async {
            // try fetch from cache initially
            val matchIdentifierList = await{matchHistoryDao.loadTwentyMatchIdenfifiers(summonerId,roleString, laneString, champId, startingPoint)}
            if (matchIdentifierList.isEmpty()) {
                val url = network.getUrlForRegion(region)
                val matchHistoryService = retrofitFactory.produceRetrofitInterface(MatchHistoryService::class.java, url)
                val result: Result<ArrayList<MatchIdentifier>> = await{matchHistoryService.fetchMatchesWithChamp(role, champId, startingPoint,summonerId).execute().body()!!}
                await{cacheMatchList(result.data)}
                presenter.onMatchListLoadedSuccessfully(result.data, result.resultCode)
            } else {
                presenter.onMatchListLoadedSuccessfully(matchIdentifierList as java.util.ArrayList<MatchIdentifier>,200)
            }
        }.onError {
            presenter.showMessage("Offline - check you internet connection")
        }
    }

//    fun loadMatches

    private fun cacheMatchList(data: ArrayList<MatchIdentifier>) {
        for (matchIdentifier in data) {
            matchHistoryDao.createMatchHistoryIdentifier(matchIdentifier)
        }
    }

    fun produceTimestamp(timestamp: Long): String {
        val oneSecond : Long = 1000
        val oneMinute : Long = 60000
        val oneHour : Long = oneMinute * 60
        val oneDay : Long = oneHour * 24
        val oneWeek : Long = oneDay * 7

        val currentTimeMillis = System.currentTimeMillis()
        val diff : Long = currentTimeMillis - timestamp
        if (diff < oneMinute) {
            val numberOfSeconds : Int = (diff / oneSecond).toInt()
            if (numberOfSeconds > 1) {
                return "$numberOfSeconds seconds ago"
            }
            return "$numberOfSeconds second ago"
        }

        if (diff >= oneMinute && diff < oneHour) {
            // convert to minutes case
            val numberOfMinutes : Int = (diff / oneMinute).toInt()
            if (numberOfMinutes > 1) {
                return "$numberOfMinutes minutes ago"
            }
            return "$numberOfMinutes minute ago"

        }
        else if (diff <oneDay) {
            val numberOfHours : Int = (diff / oneHour).toInt()
            if (numberOfHours > 1) {
                return "$numberOfHours hours ago"
            }
            return "$numberOfHours hour ago"
        } else if (diff < oneWeek) {
            val numberOfDays : Int = (diff / oneDay).toInt()
            if (numberOfDays > 1) {
                return "$numberOfDays days ago"
            }
            return "$numberOfDays day ago"
        } else {
            val numberOfWeeks : Int = (diff / oneWeek).toInt()
            if (numberOfWeeks > 1) {
                return "$numberOfWeeks weeks ago"
            }
            return "$numberOfWeeks week ago"
        }
    }

    fun loadChampName(champId: Long,
                      cardViewContract: MatchHistoryCardViewContract,
                      summonerId : String,
                      win : Boolean) {
        async {
            val champ = await { champDao.loadChamp(champId.toString(), summonerId) }
            if (champ != null) {
                cardViewContract.setChamp(champ.title)
            }
        }
    }
}
